### ISSUE TITLE
[Feature]: Have NCO detect all errors on first run

### ISSUE THREAD

**Issue by forsyth2** [E3SM-Project/zppy#535]
### How will this affect the next version number?

Small improvement (increment PATCH version)

### Is your feature request related to a problem?

When running the `climo` task on multiple variables, only the first error is reported. I.e., in #533, it took 16 runs to find out 14 variables weren't supported on `climo` whereas it only took 1 run to find out 10 variables weren't supported on `ts`.

### Describe the solution you'd like

The `climo` task should report all errors in variables rather than simply stopping after encountering one error.

**Comment by chengzhuzhang** [comment]
One potential solution is to use the fault tolerant variable list, suggested by @czender 
as mentioned here https://github.com/E3SM-Project/zppy/pull/533#issuecomment-1828872420. 

Or maybe @czender can advice if there are alternatives here.

**Comment by forsyth2** [comment]
Yes the `^var$` solution may be best. But my understanding is that then the `climo` and `ts` tasks will silently fail -- i.e. it will look like they will pass but they're really missing variables. If that's ok, I suppose I could change the code to automatically add those prefix and suffix character to every variable in the list.

**Comment by czender** [comment]
I am coming up to speed on what you are discussing. After reading #533, my impression is that timeseries generation is working as desired (i.e., failing quickly when non-present variables are requested), and that climo generation is not because it does not fail quickly enough when non-present variables are requested. Is this the essence of the problem? If not, please elaborate.

Also, why are these tasks being invoked with non-present variable names in the first place?

Once I understand both of these issues, I can be more helpful :)

**Comment by forsyth2** [comment]
> Is this the essence of the problem?

Yes, that's correct.

> why are these tasks being invoked with non-present variable names in the first place?

`zppy` was largely developed with Water Cycle in mind. We are now trying to provide Land support. This requires a completely different set of variables than what we have typically developed/tested `zppy` with.

**Comment by czender** [comment]
OK. Say `ncclimo`, in climo mode, is invoked with `-v present1,notpresent1,present2,notpresent2,...`. What is the desired behavior? to fail when processing (e.g., averaging) `notpresent1`? or to fail before processing `notpresent1`? or...? And did you know that invoking `ncclimo` without `-v varlist` causes it to climatologize all variables in the file, thus obviating the need to manage ever-changing lists of variables in `zppy`? Is that a potential solution? (It can also be invoked to exclude a list of specified variables with `-x -v varlist`)

**Comment by forsyth2** [comment]
> What is the desired behavior?

I'm thinking it would process all available variables and then fail listing "notpresent1, notpresent2 not found". Currently, it would fail with "notpresent1 not found" and the user has no idea that `notpresent2` is also not present.

> climatologize all variables in the file

I believe the reason `zppy` supports a specific `vars` list is that users usually don't want or need every single variable from a file.

**Comment by czender** [comment]
It is possible to implement this desired behavior with a change to `ncclimo`. I could interrogate the first file to see whether the varlist contains any non-present variables, and if it does, to remove those from the list, then proceed with the climatology generation with only the present variables, then fail at the end (after climatologizing the present variables), with a list of the non-present variables. Is that what y'all want? Or perhaps you want it to succeed and print only an informational message that some of the requested variables were not present? I would be happy to implement it either way.

Note that comparing the requested `varlist` to the contents of the first file and making adjustments to varlist based on the results could be done in `zppy` instead of `ncclimo`. Have you considered whether this functionality belongs there instead?

**Comment by chengzhuzhang** [comment]
I think for monthly climatology, it is not a bad idea to just include all variables, especially for v3, the output variables are much reduced.

**Comment by forsyth2** [comment]
> not a bad idea to just include all variables (as default)

Alright, we can do that as a default. But what do we want the behavior/implementation to be when users specifically set `vars`?

**Comment by czender** [comment]
I leave the desired behavior up to you two. FWIW, my "philosophy" is that it's more dangerous to fix user input errors than it is to just fail so that users are forced to learn what's in their datasets. Otherwise they'll specify non-present variables and then be surprised when those variables to not appear in successful climatologies. I think we all want to avoid that situation, that Ryan called silently ignoring bad input. This is a grey area, though, in that I think Ryan's original point was that it's currently too difficult for users to learn what's in the dataset. So I agree that some form of automated `varlist` checker may be useful. Just LMK if you want me to alter `ncclimo` to do that.

**Comment by chengzhuzhang** [comment]
> > not a bad idea to just include all variables (as default)
> 
> Alright, we can do that as a default. But what do we want the behavior/implementation to be when users specifically set `vars`?

Good question, if a user wants to get the specific set of `vars` instead of all the variables, I would assume they knows the dataset well...In any case i think a `varlist` checker will be useful. And it would be good to have it in zppy so that it can be used as needed in different places.

**Comment by rljacob** [comment]
I agree the logic to check what variables zppy hands ncclimo should live in zppy.

**Comment by rljacob** [comment]
Isn't zppy written in python?  Just open the file with python netcdf lib and query for the variables.  Don't rely on external programs from NCO.

**Comment by xylar** [comment]
I couldn't agree more.  This is something for `xarray`.

**Comment by czender** [comment]
Just remember that a valid `varlist` for `ncclimo` contains only variables that can be climatologized, i.e., rank >= 2, with at least one spatial and exactly one temporal dimension (i.e., no multi-dimensional auxiliary coordinate variables or bounds variables).
